cpp/src/arrow/array.cc

```diff
@@ -393,6 +393,71 @@ UnionArray::UnionArray(const std::shared_ptr<DataType>& type, int64_t length,
   SetData(internal_data);
 }
 
+std::shared_ptr<DataType> MakeUnionArrayType(
```

Comment:
Would this be useful as an alternate version of `arrow::union_`? Maybe with the same API (the mode as the last argument) but omitting the type codes